libkineto/src/RoctracerActivityInterface.cpp

```diff
@@ -0,0 +1,554 @@
+ * All rights reserved.
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "RoctracerActivityInterface.h"
+
+#include <cstring>
+#include <chrono>
+#include <time.h>
+
+#include "output_base.h"
+#include "ThreadUtil.h"
+
+typedef uint64_t timestamp_t;
+
+static timestamp_t timespec_to_ns(const timespec& time) {
+    return ((timestamp_t)time.tv_sec * 1000000000) + time.tv_nsec;
+  }
+
+using namespace std::chrono;
+
+namespace KINETO_NAMESPACE {
+
+constexpr size_t kBufSize(2 * 1024 * 1024);
+
+RoctracerActivityInterface& RoctracerActivityInterface::singleton() {
+  static RoctracerActivityInterface instance;
+  return instance;
+}
+
+RoctracerActivityInterface::RoctracerActivityInterface() {
+  gpuTraceBuffers_ = std::make_unique<std::list<RoctracerActivityBuffer>>();
+}
+
+RoctracerActivityInterface::~RoctracerActivityInterface() {
+  disableActivities(std::set<ActivityType>());
+  endTracing();
+}
+
+void RoctracerActivityInterface::pushCorrelationID(int id, CorrelationFlowType type) {
+#ifdef HAS_ROCTRACER
+  if (!singleton().externalCorrelationEnabled_) {
+    return;
+  }
+  // placeholder
+#endif
+}
+
+void RoctracerActivityInterface::popCorrelationID(CorrelationFlowType type) {
+#ifdef HAS_ROCTRACER
+  if (!singleton().externalCorrelationEnabled_) {
+    return;
+  }
+  // placeholder
+#endif
+}
+
+void RoctracerActivityInterface::setMaxBufferSize(int size) {
+  maxGpuBufferCount_ = 1 + size / kBufSize;
+}
+
+int RoctracerActivityInterface::processActivities(
+    ActivityLogger& logger) {
+  // Find offset to map from monotonic clock to system clock.
+  // This will break time-ordering of events but is status quo.
+
+  timespec t0, t1, t00;
+  clock_gettime(CLOCK_REALTIME, &t0);
+  clock_gettime(CLOCK_MONOTONIC, &t1);
+  clock_gettime(CLOCK_REALTIME, &t00);
+
+  const timestamp_t toffset = (timespec_to_ns(t0) >> 1) + (timespec_to_ns(t00) >> 1) - timespec_to_ns(t1);
+
+  int count = 0;
+
+  // Basic Api calls
+
+  for (auto &item : rows_) {
+    GenericTraceActivity a;
+    a.startTime = (item.begin + toffset) / 1000;
+    a.endTime = (item.end + toffset) / 1000;
+    a.id = item.id;
+    a.device = item.pid;
+    a.resource = item.tid;
+    a.activityType = ActivityType::CUDA_RUNTIME;
+    a.activityName = std::string(roctracer_op_string(ACTIVITY_DOMAIN_HIP_API, item.cid, 0));
+
+    logger.handleGenericActivity(a);
+    ++count;
+  }
+
+  // Malloc/Free calls
+  for (auto &item : mallocRows_) {
+    GenericTraceActivity a;
+    a.startTime = (item.begin + toffset) / 1000;
+    a.endTime = (item.end + toffset) / 1000;
+    a.id = item.id;
+    a.device = item.pid;
+    a.resource = item.tid;
+    a.activityType = ActivityType::CUDA_RUNTIME;
+    a.activityName = std::string(roctracer_op_string(ACTIVITY_DOMAIN_HIP_API, item.cid, 0));
+
+    a.addMetadata("ptr", fmt::format("%p", item.ptr));
```

Comment:
I believe this will simply print "%p" ?
Think you need to use the {} and fmt::ptr syntax.

Reply:
Yeah, got copy/paste happy.  Missed a step.  Hold please.